src/renderers/dom/shared/DOMPropertyOperations.js

```diff
@@ -162,41 +147,36 @@ var DOMPropertyOperations = {
    * @param {*} value
    */
   setValueForProperty: function(node, name, value) {
-    var propertyInfo = DOMProperty.getPropertyInfo(name);
-
-    if (propertyInfo && DOMProperty.shouldSetAttribute(name, value)) {
-      var mutationMethod = propertyInfo.mutationMethod;
+    if (DOMProperty.shouldSetAttribute(name, value)) {
+      var mutationMethod = DOMProperty.getMutationMethod(name);
       if (mutationMethod) {
         mutationMethod(node, value);
-      } else if (shouldIgnoreValue(propertyInfo, value)) {
+      } else if (DOMProperty.shouldIgnoreValue(name, value)) {
         DOMPropertyOperations.deleteValueForProperty(node, name);
         return;
-      } else if (propertyInfo.mustUseProperty) {
+      } else if (DOMProperty.shouldUseProperty(name)) {
         // Contrary to `setAttribute`, object properties are properly
         // `toString`ed by IE8/9.
-        node[propertyInfo.propertyName] = value;
+        node[name] = value;
       } else {
-        var attributeName = propertyInfo.attributeName;
-        var namespace = propertyInfo.attributeNamespace;
+        var attributeName = DOMProperty.getAttributeName(name);
+        var namespace = DOMProperty.getAttributeNamespace(name);
+        var expectedType = DOMProperty.getExpectedValueType(name);
         // `setAttribute` with objects becomes only `[object]` in IE8/9,
         // ('' + value) makes it output the correct toString()-value.
         if (namespace) {
           node.setAttributeNS(namespace, attributeName, '' + value);
         } else if (
-          propertyInfo.hasBooleanValue ||
-          (propertyInfo.hasOverloadedBooleanValue && value === true)
+          expectedType === 'boolean' ||
+          (expectedType === 'overloadedBoolean' && value === true)
         ) {
           node.setAttribute(attributeName, '');
         } else {
           node.setAttribute(attributeName, '' + value);
         }
       }
     } else {
-      DOMPropertyOperations.setValueForAttribute(
-        node,
-        name,
-        DOMProperty.shouldSetAttribute(name, value) ? value : null,
```

Comment:
Doesn't this change the semantics of how unknown attributes are handled since the `propertInfo` check is removed? Now unknown attributes aren't being handled by `setValueForAttribute` so they aren't validated by `isAttributeNameSafe`, which means that invalid attributes will now cause a fatal error, e.g

```
Failed to execute 'setAttribute' on 'Element': 'foo$$bar' is not a valid attribute name.
```

Reply:
In general I found the distinction between `*ForProperty` and `*ForAttribute` very confusing because it's often lying ("properties" can use attributes).

This indeed seems like a bug, can you write a test case for it?

Reply:
@gaearon I agree it's confusing, I wouldn't think that `setValueForProperty` would be setting attributes. I pushed a commit to this branch with a failing test.

Reply:
I'd really like to get rid of this distinction if we can. It always felt a like to much indirection. At the very least, it would be nice if the property and attribute paths either had clearer separation, or we consolidated them and made sure the `isAttributeNameSafe` validation is applied.

Reply:
I think as a rule of thumb, if you have an API like `setValueForProperty` it shouldn't ever hit a code path that sets it as an attribute, and vice-versa. It's confusing that there's code like `node[name] = value` inside `shouldSetAttribute` checks. From what I can tell, the naming scheme is confusing two different types of properties: what we just call "props" and properties that we need to be set as a property (what `shouldUseProperty` checks).